### PR TITLE
Updates in Raster.reproject: fix bug, simplify code and improve tests

### PR DESCRIPTION
**Pull request by adehecq**
Related to #357 and #212.
Re-organized the way `Raster.reproject` works :
- the part that handles the logic of finding the correct output goereferences has been moved to a separate (hidden) function `_get_reproject_params`. It relies now on rasterio functions (`rio.warp.calculate_default_transform` and `rio.transform.from_bounds`) only instead of making our own maths. 
- `reproject` now takes care only of the actual reprojection and handling the data
- re-organized the logic in reproject a bit and improved comments
- improve a test in TestRaster::test_reproject regarding the floating point issue
- updated a test in TestMask::test_reproject that was not actually requiring reprojection to run
- overall, improved logic in test_reproject (testing of match reference should come last instead of first), added some missing tests (coverage increased by 0.2% 🥳) and made minor edits.
- raised an issue in rasterio https://github.com/rasterio/rasterio/issues/3010

### CODE REVIEW

**Comment by adehecq** [comment]
There is an error in tests/test_raster.py::TestMask::test_reproject[mask1].
This occurs because in that case, nothing is done during reprojection: `mask_reproj = mask.reproject()` ([here](https://github.com/GlacioHack/geoutils/blob/5dcf6efdceab23d09bb95f65f61a15553cf13ec1/tests/test_raster.py#L2789)).

Three things:
- the error is caused by a bug in rasterio, see below
- to circumvent this bug, one can add a test that if dst res, size and bounds are all none, the transform and size do not change
- the test does not make sense here. It should test that reproject works as expected, but in that specific case, reproject return a copy of self, so an actual reprojection should be run

Error in rasterio:
```
import rasterio as rio
crs = rio.CRS.from_epsg(32718)
width, height, left, right, top, bottom = (539, 618, 627175.0, 643345.0, 4852085.0, 4833545.0)
rio.warp.calculate_default_transform(crs, crs, width, height, left=left, right=right, top=top, bottom=bottom)
rio.transform.from_bounds(left, bottom, right, top, width, height)
```
Because in `calculate_default_transform`, none of the crs, res or size are changed, the outcome should be the same as in `from_bounds` but it differs by a very small number...
-> need to raise an issue on rasterio and fix the bug in reproject next week!

File: tests/test_raster.py [hunk 1]
```diff
@@ -1381,24 +1381,18 @@ def test_reproject(self, example: str) -> None:
...
+        # (we take an example that used to fail, see issue #354 and #357)
+        data = np.ones((4759, 2453))
+        transform = rio.transform.Affine(
+            24.12423878332849, 0.0, 238286.29553975424, 0.0, -24.12423878332849, 6995453.456051373
+        )
         crs = rio.CRS.from_epsg(32633)
         nodata = -9999.0
         rst = gu.Raster.from_array(data=data, transform=transform, crs=crs, nodata=nodata)
 
-        # This large grid geotransform is taken from https://cdn.proj.org/us_nga_egm08_25.tif
```
Comment (rhugonnet) [inline tests/test_raster.py:1391]: Why not leave this previous example as well?

**Review by rhugonnet — APPROVED** [review]
The functionality is so much cleaner (and robust), this is great!! :grin:  

I looked at the Coveralls output (https://coveralls.io/jobs/134992379/source_files/20142500455#L331), it looks like there are 2 cases not covered in the tests:
- Bounds + Size,
- Bounds + Res as a single value.

The rest looks like checks that are duplicated between `reproject` and `_get_reproject_params`. Should we remove them from one or the other?

File: tests/test_raster.py (hunk 1 above)
Comment (adehecq) [reply tests/test_raster.py:1391]: In your example, you created two rasters (`rst` and `rst2`), one to be reprojected on the same grid as the other, to test the option with match-reference. I find these examples quite specific because the transforms are provided. I found it easier to create only one and then reproject to given bounds and res. If this one case works, by definition the match-reference case works because now this is exactly what is used for the match-reference functionality. Is that ok?

**Comment by adehecq** [comment]
> I looked at the Coveralls output (https://coveralls.io/jobs/134992379/source_files/20142500455#L331), it looks like there are 2 cases not covered in the tests:
> 
>     * Bounds + Size,
> 
>     * Bounds + Res as a single value.
> 

Coverall is really great at showing where tests are still weak ! And I wondered when writing this PR how to see the results of coveralls. 
I improved the tests accordingly. Still waiting for the tests to run on GitHub, but hopefully, tests coverage should increase now 😄 (the most annoying thing is to always have to wait ~10 min to make sure everything runs fine...)

> 
> The rest looks like checks that are duplicated between `reproject` and `_get_reproject_params`. Should we remove them from one or the other?

Good point... I deleted the one in reproject, although it means it will be caught a bit later in the code.

File: tests/test_raster.py (hunk 1 above)
Comment (rhugonnet) [reply tests/test_raster.py:1391]: OK makes sense!
